File: frontend/src/types/project.ts
/** 案件（Project）関連の型定義。 */

export type ProjectStatus =
  | "draft"
  | "submitted"
  | "won"
  | "lost";

// Phase R-1: 案件立場
export type ProjectRole = "prime" | "sub";

export const PROJECT_STATUS_LABEL: Record<ProjectStatus, string> = {
  draft:     "作成中",
  submitted: "提出済",
  won:       "受注",
  lost:      "失注",
};

export const PROJECT_STATUS_COLOR: Record<ProjectStatus, string> = {
  draft:     "#94a3b8",
  submitted: "#3b82f6",
  won:       "#22c55e",
  lost:      "#ef4444",
};

export const PROJECT_ROLE_LABEL: Record<ProjectRole, string> = {
  prime: "元請",
  sub: "下請",
};

export const PROJECT_ROLE_COLOR: Record<ProjectRole, string> = {
  prime: "#1d4ed8",
  sub: "#ea580c",
};

export interface ProjectListItem {
  id: string;
  project_number: string;
  project_name: string;
  client_name: string | null;
  status: ProjectStatus;
  project_role: ProjectRole | null;
  project_price: number | null;
  sales_person_name: string | null;
  construction_person_name: string | null;
  created_at: string;
}

export interface ProjectListResponse {
  items: ProjectListItem[];
  total: number;
  page: number;
  per_page: number;
}

export interface ProjectCreate {
  project_name: string;
  project_number?: string;
  client_name?: string;
  project_location?: string;
  sales_person_id?: string;
  construction_person_id?: string;
  project_price?: number;
}

export interface ProjectDetail {
  id: string;
  project_number: string;
  project_name: string;
  client_name: string | null;
  client_id: string | null;
  client_site_id: string | null;
  project_location: string | null;
  status: ProjectStatus;
  project_summary: string | null;
  client_contact_company: string | null;
  client_contact_person: string | null;
  client_contact_phone: string | null;
  sales_person_id: string | null;
  sales_person_name: string | null;
  construction_person_id: string | null;
  construction_person_name: string | null;
  created_by: string;
  project_price: number | null;
  project_role: ProjectRole | null;
  period_quote_start: string | null;
  period_quote_end: string | null;
  created_at: string;
  updated_at: string;
  quote_count: number;
}

/** 案件サブナビ・ProjectPickerCard で使う軽量ヘッダ型。 */
export interface ProjectHeader {
  id: string;
  projectNumber: string;
  name: string;
  status: ProjectStatus;
  client: string;
  counts: Record<string, number>;
}

export interface ProjectUpdate {
  project_name?: string;
  project_number?: string;
  client_name?: string | null;
  client_id?: string | null;
  client_site_id?: string | null;
  project_location?: string | null;
  project_role?: ProjectRole | null;
  project_summary?: string | null;
  client_contact_company?: string | null;
  client_contact_person?: string | null;
  client_contact_phone?: string | null;
  sales_person_id?: string | null;
  construction_person_id?: string | null;
  project_price?: number | null;
  period_quote_start?: string | null;
  period_quote_end?: string | null;
}

File: backend/app/schemas/project.py
"""案件（Project）関連の Pydantic スキーマ。"""
from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field

from app.models.enums import (
    ProjectRole,
    ProjectStatus,
)


class ProjectCreate(BaseModel):
    """新規案件作成リクエスト。project_number を省略すると自動採番。"""

    project_name: str = Field(..., max_length=255)
    project_number: str | None = Field(None, max_length=20)
    client_name: str | None = Field(None, max_length=255)
    project_location: str | None = None
    sales_person_id: uuid.UUID | None = None
    construction_person_id: uuid.UUID | None = None
    project_price: float | None = None
    period_quote_start: date | None = None
    period_quote_end: date | None = None


class ProjectRoleUpdate(BaseModel):
    """案件立場（元請/下請/公共）変更リクエスト。"""
    project_role: ProjectRole


class ProjectUpdate(BaseModel):
    """案件更新リクエスト。すべてのフィールドがオプション。"""

    project_name: str | None = Field(None, max_length=255)
    project_number: str | None = Field(None, max_length=20)
    client_name: str | None = None
    client_id: uuid.UUID | None = None
    client_site_id: uuid.UUID | None = None
    project_location: str | None = None
    project_summary: str | None = None
    client_contact_company: str | None = None
    client_contact_person: str | None = None
    client_contact_phone: str | None = None
    sales_person_id: uuid.UUID | None = None
    construction_person_id: uuid.UUID | None = None
    project_price: float | None = None
    period_quote_start: date | None = None
    period_quote_end: date | None = None
    project_role: ProjectRole | None = None


class ProjectListItem(BaseModel):
    """案件一覧の1行。"""

    id: uuid.UUID
    project_number: str
    project_name: str
    client_name: str | None
    status: ProjectStatus
    project_role: ProjectRole | None = None
    project_price: float | None
    sales_person_name: str | None
    construction_person_name: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class ProjectListResponse(BaseModel):
    """案件一覧レスポンス（ページネーション付き）。"""

    items: list[ProjectListItem]
    total: int
    page: int
    per_page: int


class StatusChangeRequest(BaseModel):
    """ステータス変更リクエスト。"""

    status: ProjectStatus


class EditHistoryItem(BaseModel):
    """編集履歴の1件。"""

    id: uuid.UUID
    entity_type: str
    change_type: str
    field_changes: dict | None
    changed_by_name: str
    changed_at: datetime


class EditHistoryResponse(BaseModel):
    items: list[EditHistoryItem]
    total: int


class ProjectCounts(BaseModel):
    """案件サブナビ用の関連データ件数。"""

    estimate: int = 0
    quote: int = 0
    history: int = 0


class ProjectDetail(BaseModel):
    """案件詳細レスポンス（全フィールド＋関連件数）。"""

    id: uuid.UUID
    project_number: str
    project_name: str
    client_name: str | None
    client_id: uuid.UUID | None
    client_site_id: uuid.UUID | None
    project_location: str | None
    status: ProjectStatus
    project_summary: str | None
    client_contact_company: str | None
    client_contact_person: str | None
    client_contact_phone: str | None
    sales_person_id: uuid.UUID | None
    sales_person_name: str | None
    construction_person_id: uuid.UUID | None
    construction_person_name: str | None
    created_by: uuid.UUID
    project_price: float | None
    project_role: ProjectRole | None = None
    period_quote_start: date | None
    period_quote_end: date | None
    created_at: datetime
    updated_at: datetime
    quote_count: int
    # サブナビ用集計オブジェクト
    counts: ProjectCounts
